File: tools/bench/csr_traversal.py
"""인접 리스트(list of list) vs CSR(array.array) — XI-10 §2·§4 의 Python 쪽 수치.

시간과 메모리를 **다른 실행**에서 잰다. 한 프로세스에서 둘 다 재면 GC 와 할당자
상태가 섞이고, 계측을 켠 채로 잰 시간은 몇십 배까지 부푼다.

사용법:
    python3.13 tools/bench/csr_traversal.py time
    python3.13 tools/bench/csr_traversal.py mem-adj
    python3.13 tools/bench/csr_traversal.py mem-csr
    python3.13 tools/bench/csr_traversal.py sizeof
"""

import array
import random
import statistics
import sys
import time

V = 200000
E = 2000000  # 무방향. 인접 자리는 2E = 400만 개


def rss_mb():
    with open("/proc/self/statm") as f:
        resident = int(f.read().split()[1])
    return resident * 4096 / 1024 / 1024


def make_edges():
    rng = random.Random(20260805)
    return [(rng.randrange(V), rng.randrange(V)) for _ in range(E)]


def build_adj(es):
    adj = [[] for _ in range(V)]
    for u, v in es:
        adj[u].append(v)
        adj[v].append(u)
    return adj


def build_csr(es):
    start = array.array("i", bytes(4 * (V + 1)))
    for u, v in es:
        start[u + 1] += 1
        start[v + 1] += 1
    for i in range(1, V + 1):
        start[i] += start[i - 1]
    item = array.array("i", bytes(4 * 2 * E))
    fill = array.array("i", start[:V])
    for u, v in es:
        item[fill[u]] = v
        fill[u] += 1
        item[fill[v]] = u
        fill[v] += 1
    return start, item


def main():
    mode = sys.argv[1] if len(sys.argv) > 1 else "time"
    es = make_edges()

    if mode == "sizeof":
        # 리스트 한 칸은 객체 포인터 8바이트다. 그 포인터가 가리키는 int 객체가 또 있다.
        adj = build_adj(es)
        slots = sum(len(row) for row in adj)
        outer = sys.getsizeof(adj)
        inner = sum(sys.getsizeof(row) for row in adj)
        start, item = build_csr(es)
        print(f"인접 자리 {slots:,}개")
        print(f"  list of list : 바깥 {outer / 1e6:.1f} MB + 안쪽 {inner / 1e6:.1f} MB"
              f" = {(outer + inner) / 1e6:.1f} MB (int 객체 본체는 여기 안 들어 있다)")
        print(f"  int 객체 하나 {sys.getsizeof(V - 1)} 바이트 × 서로 다른 값 최대 {V:,}개")
        print(f"  array.array  : start {sys.getsizeof(start) / 1e6:.1f} MB"
              f" + item {sys.getsizeof(item) / 1e6:.1f} MB"
              f" = {(sys.getsizeof(start) + sys.getsizeof(item)) / 1e6:.1f} MB")
        return
    if mode == "mem-adj":
        before = rss_mb()
        adj = build_adj(es)
        print(f"인접 리스트 상주 메모리 증가: {rss_mb() - before:.0f} MB"
              f" (인접 자리 {sum(len(r) for r in adj):,}개)")
        return
    if mode == "mem-csr":
        before = rss_mb()
        start, item = build_csr(es)
        print(f"CSR 상주 메모리 증가: {rss_mb() - before:.0f} MB"
              f" (item {len(item):,}칸)")
        return

    print(f"정점 {V:,} · 간선 {E:,}(무방향, 인접 자리 {2 * E:,}) · 5회 중앙값과 범위")
    res = {k: [] for k in ("adj_build", "csr_build", "adj_walk", "csr_walk")}
    for _ in range(5):
        t0 = time.perf_counter()
        adj = build_adj(es)
        t1 = time.perf_counter()
        s = 0
        for row in adj:
            s += sum(row)
        t2 = time.perf_counter()
        del adj

        t3 = time.perf_counter()
        start, item = build_csr(es)
        t4 = time.perf_counter()
        s2 = 0
        for u in range(V):
            s2 += sum(item[start[u]:start[u + 1]])
        t5 = time.perf_counter()
        assert s == s2, (s, s2)
        del start, item

        res["adj_build"].append((t1 - t0) * 1000)
        res["adj_walk"].append((t2 - t1) * 1000)
        res["csr_build"].append((t4 - t3) * 1000)
        res["csr_walk"].append((t5 - t4) * 1000)

    for k, label in (
        ("adj_build", "인접 리스트 구축"),
        ("csr_build", "CSR 구축"),
        ("adj_walk", "인접 리스트 순회"),
        ("csr_walk", "CSR 순회"),
    ):
        v = sorted(res[k])
        print(f"  {label:<20} {statistics.median(v):8.0f} ms  ({v[0]:.0f} ~ {v[-1]:.0f})")


if __name__ == "__main__":
    main()

File: tools/bench/csr_traversal.cpp
// 인접 리스트(vector<vector<int>>) vs CSR — XI-10 §2·§4 의 수치.
//
// 같은 그래프를 두 표현으로 들고 (1) 만드는 시간 (2) 전체 순회 시간 (3) 상주 메모리를 잰다.
// 시간과 메모리는 서로 다른 실행에서 잰다 — 한 프로세스에서 둘 다 재면 할당자 상태가 섞인다.
//
// 빌드: g++ -std=c++17 -O2 tools/bench/csr_traversal.cpp -o /tmp/csr
// 실행: /tmp/csr time   또는   /tmp/csr mem-adj   또는   /tmp/csr mem-csr

#include <algorithm>
#include <chrono>
#include <cstdint>
#include <cstdio>
#include <cstring>
#include <unistd.h>
#include <random>
#include <string>
#include <vector>
using namespace std;

const int V = 1000000;
const int E = 10000000;   // 무방향. 인접 자리는 2E = 2천만 개

static long rss_kb() {
    FILE* f = fopen("/proc/self/statm", "r");
    long size = 0, resident = 0;
    if (f) { if (fscanf(f, "%ld %ld", &size, &resident) != 2) resident = 0; fclose(f); }
    return resident * (sysconf(_SC_PAGESIZE) / 1024);
}

static vector<pair<int, int>> make_edges() {
    mt19937 rng(20260805);
    vector<pair<int, int>> es(E);
    for (int i = 0; i < E; i++) {
        int u = (int)(rng() % V), v = (int)(rng() % V);
        es[i] = {u, v};
    }
    return es;
}

static double ms(chrono::steady_clock::time_point a, chrono::steady_clock::time_point b) {
    return chrono::duration<double, milli>(b - a).count();
}

int main(int argc, char** argv) {
    string mode = argc > 1 ? argv[1] : "time";
    long base = rss_kb();
    auto es = make_edges();

    if (mode == "mem-adj") {
        long before = rss_kb();
        vector<vector<int>> adj(V);
        for (auto [u, v] : es) { adj[u].push_back(v); adj[v].push_back(u); }
        printf("인접 리스트 상주 메모리 증가: %ld MB\n", (rss_kb() - before) / 1024);
        printf("  (이론 하한 %ld MB = 안쪽 vector 헤더 24B*V + 원소 4B*2E)\n",
               ((long)V * 24 + (long)2 * E * 4) / 1024 / 1024);
        return 0;
    }
    if (mode == "mem-csr") {
        long before = rss_kb();
        vector<int> start(V + 1, 0), item(2 * E);
        for (auto [u, v] : es) { start[u + 1]++; start[v + 1]++; }
        for (int i = 1; i <= V; i++) start[i] += start[i - 1];
        vector<int> fill(start.begin(), start.end() - 1);
        for (auto [u, v] : es) { item[fill[u]++] = v; item[fill[v]++] = u; }
        printf("CSR 상주 메모리 증가: %ld MB\n", (rss_kb() - before) / 1024);
        printf("  (이론값 %ld MB = start 4B*(V+1) + item 4B*2E, 임시 fill 제외)\n",
               ((long)(V + 1) * 4 + (long)2 * E * 4) / 1024 / 1024);
        return 0;
    }

    printf("정점 %d · 간선 %d(무방향, 인접 자리 %d) · 5회 중앙값과 범위\n", V, E, 2 * E);
    // BFS·DFS 는 정점을 번호 순으로 보지 않는다. 무작위 순서 순회도 함께 잰다.
    vector<int> order(V);
    for (int i = 0; i < V; i++) order[i] = i;
    shuffle(order.begin(), order.end(), mt19937(1234));
    vector<double> ba, bc, ta, tc, ra, rc;
    long long sink = 0;
    for (int rep = 0; rep < 5; rep++) {
        auto t0 = chrono::steady_clock::now();
        vector<vector<int>> adj(V);
        for (auto [u, v] : es) { adj[u].push_back(v); adj[v].push_back(u); }
        auto t1 = chrono::steady_clock::now();

        long long s = 0;
        for (int u = 0; u < V; u++)
            for (int v : adj[u]) s += v;
        auto t2 = chrono::steady_clock::now();
        long long sr = 0;
        for (int i = 0; i < V; i++)
            for (int v : adj[order[i]]) sr += v;
        auto t2r = chrono::steady_clock::now();
        vector<vector<int>>().swap(adj);
        if (sr != s) printf("  !! 순서만 다른데 합이 다르다\n");

        auto t3 = chrono::steady_clock::now();
        vector<int> start(V + 1, 0), item(2 * E);
        for (auto [u, v] : es) { start[u + 1]++; start[v + 1]++; }
        for (int i = 1; i <= V; i++) start[i] += start[i - 1];
        {
            vector<int> fill(start.begin(), start.end() - 1);
            for (auto [u, v] : es) { item[fill[u]++] = v; item[fill[v]++] = u; }
        }
        auto t4 = chrono::steady_clock::now();

        long long s2 = 0;
        for (int u = 0; u < V; u++)
            for (int k = start[u]; k < start[u + 1]; k++) s2 += item[k];
        auto t5 = chrono::steady_clock::now();
        long long s2r = 0;
        for (int i = 0; i < V; i++) {
            int u = order[i];
            for (int k = start[u]; k < start[u + 1]; k++) s2r += item[k];
        }
        auto t5r = chrono::steady_clock::now();
        if (s2r != s2) printf("  !! 순서만 다른데 합이 다르다\n");

        if (s != s2) printf("  !! 두 표현의 합이 다르다: %lld vs %lld\n", s, s2);
        sink += s;
        ba.push_back(ms(t0, t1)); ta.push_back(ms(t1, t2)); ra.push_back(ms(t2, t2r));
        bc.push_back(ms(t3, t4)); tc.push_back(ms(t4, t5)); rc.push_back(ms(t5, t5r));
    }
    auto rep = [](const char* n, vector<double> v) {
        sort(v.begin(), v.end());
        printf("  %-26s %8.0f ms  (%.0f ~ %.0f)\n", n, v[2], v.front(), v.back());
    };
    rep("인접 리스트 구축", ba);
    rep("CSR 구축", bc);
    rep("인접 리스트 순회(번호순)", ta);
    rep("CSR 순회(번호순)", tc);
    rep("인접 리스트 순회(무작위)", ra);
    rep("CSR 순회(무작위)", rc);
    printf("  (합 대조용 %lld, 기준 RSS %ld MB)\n", sink % 7, base / 1024);
    return 0;
}
